openshift/machine-config-operator
# MCO-1100: enable RHEL entitlements in on-cluster layering

**- What I did**

This adds the capability for BuildController to use the RHEL entitlement secrets to allow cluster admins to inject RHEL content into their builds that they are entitled to receive. This also allows the injection / consumption of content into `/etc/yum.repos.d` as well as `/etc/pki/rpm-gpg`. There are a few notes about the implementation that I would like to have at a higher level:

- Because we run rootless Buildah, we're more prone to running into SELinux complications. This makes it more difficult  to directly mount the contents of `/etc/yum.repos.d`, `/etc/pki/entitlement`, and `/etc/pki/rpm-gpg` directly into the build context. With that in mind, we copy everything into a series of temp directories first, and then mount those temp directories into the build context as a volume.
- We also create an `emptyDir` which is mounted into the build pod at `/home/build/.local/share/containers`. It is unclear why this is necessary, but as mentioned before, I suspect that this is due to SELinux issues.
- The e2e test suite now has the capability to stream the container logs from the build pod to the filesystem as there is useful information contained within those logs if the e2e test fails. In OpenShift CI, this location will be determined by the `ARTIFACT_DIR` env var. If this env var is not present, it will default the current directory.

**- How to verify it**

Automated verification:

1. Bring up a cluster where the secret `etc-pki-entitlement` exists in the `openshift-config-managed` namespace. If this secret is not present, `TestEntitledBuilds` and `TestEntitledBuildsRollsOutImage` will be skipped.
2. Ensure that the OnClusterBuild feature-gate is enabled. The test suite will fail immediately if the feature-gate is not enabled.
3. Run the tech preview e2e test suite: `$ go test -count=1 -v ./test/e2e-techpreview/...`

(Note: Because we have not landed https://github.com/openshift/machine-config-operator/pull/4284, the cleanup / teardown will delete the node and its underlying machine, causing the Machine API to provision a replacement node.)

Semi-manual verification:

1. Download / install [v0.0.14](https://github.com/cheesesashimi/zacks-openshift-helpers/releases/tag/v0.0.14) (or newer) of my OpenShift helpers on your local machine.
2. Bring up a cluster where the secret `etc-pki-entitlement` exists in the `openshift-config-managed` namespace.
3. Create a Dockerfile on your local machine that contains the following content:
```Dockerfile
FROM configs AS final

RUN rm -rf /etc/rhsm-host && \
  rpm-ostree install buildah && \
  ln -s /run/secrets/rhsm /etc/rhsm-host && \
  ostree container commit
```
5. With my `onclustertesting` helper in your `$PATH`, run the following: `$ onclustertesting setup in-cluster-registry --enable-featuregate --pool=layered --custom-dockerfile=./path/to/the/Dockerfile`
6. If you have not previously enabled the featuregate, my helper will enable it for you. It will cause a new MachineConfig to be created and rolled out to all of the nodes, so the build might not begin immediately. Using this flag is idempotent.
7. Watch for the `machine-os-builder` pod to start. Shortly afterward, the build pod should start. It should complete without any errors.

**- Description for the changelog**
Enables RHEL entitlements in on-cluster layering
